tools/dockerfile-add-scanner/cli/scan.go

```diff
@@ -0,0 +1,160 @@
+package cli
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/moby/buildkit/frontend/dockerfile/dockerfile2llb"
+	"github.com/moby/buildkit/solver/pb"
+	ocispecs "github.com/opencontainers/image-spec/specs-go/v1"
+	log "github.com/sirupsen/logrus"
+	spdxjson "github.com/spdx/tools-golang/json"
+	spdxcommon "github.com/spdx/tools-golang/spdx/common"
+	spdx "github.com/spdx/tools-golang/spdx/v2_3"
+	spdxtv "github.com/spdx/tools-golang/tvsaver"
+	"github.com/spf13/cobra"
+)
+
+const (
+	defaultNamespace = "https://github.com/lf-edge/eve/spdx"
+	creator          = "https://github.com/lf-edge/eve/tools/dockerfile-add-scanner"
+)
+
+func scanCmd() *cobra.Command {
+	var (
+		outputFormat string
+		namespace    string
+	)
+	cmd := &cobra.Command{
+		Use:   "scan",
+		Short: "Scan Dockerfiles for ADD commands",
+		Long: `Scan Dockerfiles for ADD commands and print the URLs to stdout.
+	Can scan multiple at once. Output can be in list, spdx or spdx-json formats.
+`,
+		Example: `dockerfile-add-scanner scan <dockerfile1> <dockerfile2> ... <dockerfileN>`,
+		Args:    cobra.MinimumNArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			var allUrls []string
+			for _, dockerfile := range args {
+				log.Debugf("Processing %s", dockerfile)
+				urls, err := scan(dockerfile)
+				if err != nil {
+					log.Fatalf("Error scanning %s: %v", dockerfile, err)
+				}
+				allUrls = append(allUrls, urls...)
+			}
+			switch outputFormat {
+			case "list":
+				fmt.Println(strings.Join(allUrls, "\n"))
+			case "spdx":
+				sbom, err := buildSbom(allUrls, namespace, creator)
+				if err != nil {
+					return err
+				}
+				return spdxtv.Save2_3(sbom, os.Stdout)
+			case "spdx-json":
+				sbom, err := buildSbom(allUrls, namespace, creator)
+				if err != nil {
+					return err
+				}
+				return spdxjson.Save2_3(sbom, os.Stdout)
+			default:
+				return fmt.Errorf("unknown output format %s", outputFormat)
+			}
+			return nil
+		},
+	}
+
+	cmd.Flags().StringVar(&outputFormat, "format", "list", "Output format: list, spdx, spdx-json")
+	cmd.Flags().StringVar(&namespace, "namespace", defaultNamespace, "document namespace to use for spdx output formats, will have a UUID appended")
+
+	return cmd
+}
+
+func scan(dockerfile string) ([]string, error) {
+	var urls []string
+	f, err := os.Open(dockerfile)
+	if err != nil {
+		return nil, fmt.Errorf("error opening dockerfile %s: %v", dockerfile, err)
+	}
+	defer f.Close()
+	data, err := io.ReadAll(f)
+	if err != nil {
+		return nil, fmt.Errorf("error reading dockerfile at %s: %v", dockerfile, err)
+	}
+
+	ctx := context.TODO()
+	state, _, _, err := dockerfile2llb.Dockerfile2LLB(ctx, data, dockerfile2llb.ConvertOpt{
+		TargetPlatform: &ocispecs.Platform{Architecture: "amd64", OS: "linux"},
```

Comment:
At first blush, I was thinking that we do not care, since we are just scanning the dockerfile. Then I realized that if it is in a stage that is not executed for that arch, we will include it when we shouldn't. So I can add that as a flag, defaulting to the current platform.

Reply:
Done